buffer_pool/buffer.go

```diff
@@ -0,0 +1,246 @@
+package buffer_pool
+
+import (
+	"errors"
+	"io"
+
+	"github.com/akitasoftware/akita-libs/memview"
+)
+
+// A variable-sized buffer whose backing storage is drawn from a fixed-sized
+// pool. Clients must return the backing storage to the pool by calling Release.
+//
+// Based on bytes.Buffer.
+type Buffer interface {
+	// Returns a MemView of length Len() that holds the unread portion of the
+	// buffer. The MemView is valid for use only until the next buffer
+	// modification (that is, only until the next call to a method like Read,
+	// Write, Reset, or Truncate). The MemView aliases the buffer content at least
+	// until the next buffer modification, so changes to the MemView will affect
+	// the result of future reads.
+	Bytes() memview.MemView
+
+	// Returns the number of bytes of the unread portion of the buffer; Len() ==
+	// Bytes().Len().
+	Len() int
+
+	// Empties the buffer. An alias for Release.
+	Reset()
+
+	// Empties the buffer and returns its underlying storage to the pool.
+	Release()
+
+	// Write(p) appends the contents of the slice p to the buffer, obtaining
+	// additional storage from the pool as needed.
+	//
+	// Returns the number of bytes written from p and EmptyPool if the write
+	// stopped early.
+	io.Writer
+
+	io.ReaderFrom
+}
+
+var ErrEmptyPool = errors.New("buffer_pool.Buffer: pool is empty")
+var errNegativeRead = errors.New("buffer_pool.Buffer: reader returned negative count from Read")
+
+type buffer struct {
+	pool bufferPool
+
+	// Contents of the buffer start at chunks[0][readOffset] (inclusive) and end
+	// at chunks[len(chunks)-1][writeOffset] (exclusive).
+	//
+	// Invariants:
+	//   - this is empty when the buffer is empty.
+	//   - all elements have length and capacity pool.chunkSize_bytes.
+	chunks [][]byte
+
+	// Contents of the buffer start at chunks[0][readOffset] (inclusive). This is
+	// where to start reading from.
+	//
+	// Invariants:
+	//   - readOffset == 0 when len(chunks) == 0.
+	//   - readOffset < pool.chunkSize_bytes when len(chunks) > 0.
+	//   - readOffset < writeOffset when len(chunks) == 1.
```

Comment:
I think you should express these in code (although I see there is no unit test in which to call it.)

While testing in our own infrastructure we could enable this invariant-checking as an assertion at the exit from ReadFrom and Write.

Reply:
Added in cf20c0a. Unit tests coming soon!

Reply:
Unit tests in 1caff2e (refactored in 5148008).

Reply:
Invariant-checking enabled during dogfooding in akitasoftware/akita-cli@b2dc2550cffe3022c7bc035b49aed3aa337322a2 as part of akitasoftware/akita-cli#147.